File: benchmarks/17-fft-and-signal/fft_and_signal.m
(* Experiment 17 -- FFT, convolution, correlation.
   MEASURES src/fourier.c and src/convolutions.c.

   READ THE BUILD WARNING FIRST.  Without FFTW the makefile substitutes "a naive
   O(n^2) fallback" and every row here measures that instead: on this tree a
   32768-point transform was 1050 ms without FFTW and 1.0 ms with it -- a factor
   of a thousand that has nothing to do with src/fourier.c's own quality.

   Power-of-two AND non-power-of-two sizes are both here, because a mixed-radix
   size is exactly where a weak FFT falls back to something quadratic. *)

Get["../harness.m"];
Get["../data.m"];

require[{"Fourier", "InverseFourier", "ListConvolve", "ListCorrelate",
         "FourierDCT", "PeriodogramArray"}];

(* Wolfram's Fourier normalises by 1/Sqrt[n]; the checks account for that. *)
small = {1., 2., 3., 4., 5., 6., 7., 8.};

d1 = rand01[{262144}];              (* 2^18: the friendly case *)
bench["Fourier 2^18 (262144)", Fourier[d1];];
check["Fourier 2^18 (262144)", Round[10^6 Re[First[Fourier[small]]]]];

d2 = rand01[{1200000}];             (* 2^7 * 3 * 5^5: mixed radix *)
bench["Fourier 1200000 (mixed radix)", Fourier[d2];];
check["Fourier 1200000 (mixed radix)", Round[10^6 Re[Last[Fourier[small]]]]];

d3 = rand01[{262143}];              (* 3^3 * 7 * ... : an awkward size *)
bench["Fourier 262143 (awkward size)", Fourier[d3];];
check["Fourier 262143 (awkward size)",
  Round[10^6 Abs[Im[Part[Fourier[small], 2]]]]];

bench["InverseFourier 2^18", InverseFourier[d1];];
check["InverseFourier 2^18",
  Round[10^6 Re[First[InverseFourier[Fourier[small]]]]]];

(* Convolution of two length-100000 sequences: the FFT path if there is one. *)
a = rand01[{100000}]; b = rand01[{2048}];
bench["ListConvolve 100000 x 2048", ListConvolve[b, a];];
check["ListConvolve 100000 x 2048",
  Round[10^6 First[ListConvolve[{1., 2.}, {3., 4., 5.}]]]];

bench["ListCorrelate 100000 x 2048", ListCorrelate[b, a];];
check["ListCorrelate 100000 x 2048",
  Round[10^6 First[ListCorrelate[{1., 2.}, {3., 4., 5.}]]]];

File: benchmarks/17-fft-and-signal/fft_and_signal.py
#!/usr/bin/env python3
"""Experiment 17 -- FFT, convolution, correlation (scipy.fft / numpy column).

Same six kernels as ``fft_and_signal.m``, same order and sizes.

Power-of-two AND non-power-of-two sizes are both here, because a mixed-radix
size is exactly where a weak FFT falls back to something quadratic.  numpy's
pocketfft handles all three well, so the spread across these rows is a property
of the CAS columns.

NORMALISATION.  Wolfram's ``Fourier`` divides by sqrt(n); numpy's ``fft`` does
not.  Only the CHECK values are affected -- the timing is the same transform --
so each check divides by sqrt(8) for the 8-point reference sequence.
"""

import sys, os; sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import math

import numpy as np

from harness import bench, check, require, seed

require(["numpy.fft:fft", "numpy.fft:ifft", "numpy:convolve",
         "numpy:correlate", "scipy.signal:fftconvolve"])

seed()
small = np.array([1., 2., 3., 4., 5., 6., 7., 8.])
S = math.sqrt(8.0)


def r6(x):
    return int(math.floor(float(x) * 1e6 + 0.5))


d1 = np.random.random(262144)
bench("Fourier 2^18 (262144)", lambda: np.fft.fft(d1))
check("Fourier 2^18 (262144)", r6(np.fft.fft(small)[0].real / S))

d2 = np.random.random(1200000)
bench("Fourier 1200000 (mixed radix)", lambda: np.fft.fft(d2))
check("Fourier 1200000 (mixed radix)", r6(np.fft.fft(small)[-1].real / S))

d3 = np.random.random(262143)
bench("Fourier 262143 (awkward size)", lambda: np.fft.fft(d3))
# Wolfram uses e^{+2 pi i}, numpy e^{-2 pi i}: the imaginary part is conjugated,
# so compare |Im| and the row measures the transform rather than the convention.
check("Fourier 262143 (awkward size)", r6(abs(np.fft.fft(small)[1].imag) / S))

# Wolfram's InverseFourier also carries a 1/sqrt(n); the round trip is identity
# in both systems, which is what the check pins.
bench("InverseFourier 2^18", lambda: np.fft.ifft(d1))
check("InverseFourier 2^18", r6(np.fft.ifft(np.fft.fft(small))[0].real))

a = np.random.random(100000)
b = np.random.random(2048)
# Wolfram's ListConvolve[b, a] returns the VALID (no-padding) region.
bench("ListConvolve 100000 x 2048", lambda: np.convolve(a, b, mode="valid"))
# ListConvolve[{1,2},{3,4,5}] == {10, 13}: the kernel is applied reversed.
check("ListConvolve 100000 x 2048",
      r6(np.convolve(np.array([3., 4., 5.]), np.array([1., 2.]),
                     mode="valid")[0]))

bench("ListCorrelate 100000 x 2048", lambda: np.correlate(a, b, mode="valid"))
check("ListCorrelate 100000 x 2048",
      r6(np.correlate(np.array([3., 4., 5.]), np.array([1., 2.]),
                      mode="valid")[0]))
